src/main/utility/mod.rs

```diff
@@ -19,23 +19,75 @@ pub mod stream_len;
 pub mod syscall;
 pub mod time;
 
+use crate::{core::worker::Worker, host::host::HostId};
+
 /// A type that allows us to make a pointer Send + Sync since there is no way
 /// to add these traits to the pointer itself.
 #[derive(Clone, Copy, Debug)]
-pub struct SyncSendPointer<T>(pub *mut T);
+pub struct SyncSendPointer<T>(*mut T);
 
 unsafe impl<T> Send for SyncSendPointer<T> {}
 unsafe impl<T> Sync for SyncSendPointer<T> {}
 
 impl<T> SyncSendPointer<T> {
-    /// Get the pointer.
+    /// SAFETY: The object pointed to by `ptr` must actually be `Send` and
+    /// `Sync`, or else not subsequently used in contexts where it matters.
+    pub unsafe fn new(ptr: *mut T) -> Self {
+        Self(ptr)
+    }
+
     pub fn ptr(&self) -> *mut T {
         self.0
     }
+}
+
+/// A type that allows us to make a pointer Send + Sync since there is no way
+/// to add these traits to the pointer itself.
+#[derive(Clone, Copy, Debug)]
+pub struct HostTreePointer<T> {
+    #[cfg(debug_assertions)]
+    host_id: HostId,
+    ptr: *mut T,
+}
+
+unsafe impl<T> Send for HostTreePointer<T> {}
+unsafe impl<T> Sync for HostTreePointer<T> {}
+
+impl<T> HostTreePointer<T> {
+    /// SAFETY: The object pointed to by `ptr` must either actually be `Send`
+    /// and `Sync`, or else not subsequently used in contexts where it matters.
+    ///
+    /// In Shadow typically the object itself may not be `Send` or `Sync`, but
+    /// safety is guaranteed by the pointer living within a single Host's
+    /// "object tree", and accessed with that Host's lock held.
+    pub unsafe fn new_for_host(host_id: HostId, ptr: *mut T) -> Self {
+        // Suppress "never read" warning in release builds.
+        let _ = host_id;
+        Self {
+            #[cfg(debug_assertions)]
+            host_id,
+            ptr,
+        }
+    }
+
+    /// Create a pointer for the current host.
+    ///
+    /// SAFETY: see `new_for_host`.
+    pub unsafe fn new(ptr: *mut T) -> Self {
+        let host_id = Worker::with_active_host_info(|i| i.id);
+        unsafe { Self::new_for_host(host_id.unwrap(), ptr) }
+    }
 
-    /// Get a mutable reference to the pointer.
-    pub fn ptr_ref(&mut self) -> &mut *mut T {
-        &mut self.0
+    /// Get the pointer.
+    ///
+    /// SAFETY: Pointer must only be dereferenced with the Host lock held, in
+    /// addition to the normal safety requirements for dereferencing a pointer.
+    pub unsafe fn ptr(&self) -> *mut T {
+        #[cfg(debug_assertions)]
+        Worker::with_active_host_info(|i| {
+            debug_assert_eq!(self.host_id, i.id);
+        });
+        self.ptr
     }
```

Comment:
Why make this a debug-only check? Rust types that check safety always run in release mode as well, so I don't see a reason not to here. It would also allow us to make `ptr()` safe. Or what about making it return a `Result<*mut T, ()>` that the calling code needs to `unwrap()`?

Reply:
No longer debug-only.